### ISSUE TITLE
OLD iterators Concept Exercise

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2367]
This issue describes how to implement  the  `iterators` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/creation/use of `iterators`  in Python. 

## Learning objectives

Learn more about how `looping` / `iteration` and the `iterator protocol` work in Python, and the options available for creating & customizing `iterators`.

- How `loops` in Python implement the `iterator protocol`
-  `iteration`  "under the covers" 
- How `iterables` are defined
- Understand and use the built-in `iter()` to return an `iterator` 
- Create/use one or more custom `iterators` via defining a `class`
- Understand how `class` -created iterators relate to `generators`, and the pros and cons of each approach

  

## Out of scope

- `classes`  & `class customization` beyond the use of the `iterator` dunder methods and `iterator protocol`
- `class-inheritance` beyond what is needed to customize `iteration` for a `container type` or other `class`
- `comprehensions`
- `comprehensions` in `lambdas`
- `coroutines` 
- `decorators`
-  `functools` and related `map()`, `filter()` and `functools.reduce()`
- `generators` in-depth as a specific form of iterators (_these have their own exercise_)
- `higher-order functions`
- `lambdas`
- using an `assignment expression` or "walrus" operator (`:=`) 
- class decorators 
- `enums`

  

## Concepts

- `loops`
- `iteration`
- `iterators` & `iterator types`
- `iterator protocol`
- `iterables`
- `enumeration`

  

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `classes`
- `comparisons`
- `rich-comparisons`
- `decorators`
- `descriptors`
- `dicts`
- `dict-methods`
- `functions`
- `higher-order-functions`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

  

## Resources to refer to

- [Python Docs: iterator definition](https://docs.python.org/3/glossary.html#term-iterator) | [Python Docs: iterator types](https://docs.python.org/3/library/stdtypes.html#iterator-types) 
  -  [Python Docs: iterable](https://docs.python.org/3/glossary.html#term-iterable) 
  - [Python Tutorial: iterators](https://docs.python.org/3/tutorial/classes.html#iterators) | [Python Functional Programming HOWTO: Iterators](https://docs.python.org/3/howto/functional.html#iterators)
  -  [Python Docs: sequence](https://docs.python.org/3/glossary.html#term-sequence)
  -  [Python Docs: the `for` statement](https://docs.python.org/3/tutorial/controlflow.html#for-statements)
  - [Trey Hunner: The Iterator Protocol: How `for` Loops Work in Python](https://treyhunner.com/2016/12/python-iterator-protocol-how-for-loops-work/)
  - [Dan Bader: Python Iterators - A Step-by-Step Introduction](https://dbader.org/blog/python-iterators)
- [`iter()`](https://docs.python.org/3/library/functions.html#iter)
- [Python Docs: eumulating container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types)
  - [Trey Hunner: How to make an `iterator` in Python](https://treyhunner.com/2018/06/how-to-make-an-iterator-in-python/)
  - [Python Patterns Guide: Gang of Four `iterator` Pattern](https://python-patterns.guide/gang-of-four/iterator/)
- [Python Docs C-API: Iterator Protocol](https://docs.python.org/3/c-api/iter.html)
- [Python Docs C-API: Iterator Objects](https://docs.python.org/3/c-api/iterator.html#iterator-objects)

  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is currently no issue or files for the concept.  **They are TBD.**

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  - This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `087e3eec-33f4-4b7e-8335-a5ee9ddda9f3`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Closing for re-write
